projects/angular/src/forms/datepicker/date-single-input.ts

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright (c) 2016-2024 Broadcom. All Rights Reserved.
+ * The term "Broadcom" refers to Broadcom Inc. and/or its subsidiaries.
+ * This software is released under MIT license.
+ * The full license information can be found in LICENSE in the root directory of this project.
+ */
+
+import { AfterViewInit, Directive, EventEmitter, HostListener, Input, Output } from '@angular/core';
+import { of } from 'rxjs';
+import { filter, switchMap } from 'rxjs/operators';
+
+import { ClrDateInputBase } from './date-input';
+import { DayModel } from './model/day.model';
+import { DatepickerFocusService } from './providers/datepicker-focus.service';
+import { datesAreEqual } from './utils/date-utils';
+
+@Directive({
+  selector: '[clrDate]',
+  host: {
+    '[class.clr-input]': 'true',
+  },
+  providers: [DatepickerFocusService],
+})
+export class ClrDateInput extends ClrDateInputBase implements AfterViewInit {
+  @Output('clrDateChange') dateChange = new EventEmitter<Date>(false);
+
+  private initialClrDateInputValue: Date;
+  private previousDateChange: Date;
+
+  @Input('clrDate')
+  set date(date: Date | string) {
+    if (date && typeof date === 'string') {
+      date = new Date(date);
+    }
+    if (this.previousDateChange !== date) {
+      this.updateDate(this.getValidDateValueFromDate(date as Date));
+    }
+
+    if (!this.initialClrDateInputValue) {
+      this.initialClrDateInputValue = date as Date;
+    }
+  }
+
+  @Input()
+  set min(dateString: string) {
+    this.dateIOService.setMinDate(dateString);
+    this.triggerControlValidation();
+  }
+
+  @Input()
+  set max(dateString: string) {
+    this.dateIOService.setMaxDate(dateString);
+    this.triggerControlValidation();
+  }
+
+  @HostListener('change', ['$event.target'])
+  onValueChange(target: HTMLInputElement) {
+    const validDateValue = this.dateIOService.getDateValueFromDateString(target.value);
+    if (this.usingClarityDatepicker() && validDateValue) {
+      this.updateDate(validDateValue, true);
+    } else if (this.usingNativeDatepicker()) {
+      const [year, month, day] = target.value.split('-');
+      this.updateDate(new Date(+year, +month - 1, +day), true);
+    } else {
+      this.emitDateOutput(null);
+    }
+  }
```

Comment:
This method is repeated in all derived classes. It can be moved back to the base class. See my other comments about adding `updateDate` as an abstract method and moving `emitDateOutput` back to the base class as well.

Reply:
I will provide an example commit that moves all of the duplicated code back to the base class.